### PR TITLE
feat: clarify confusion matrix confidence threshold during validation

### LINKED ISSUE #25724: Model evaluation yields incoherent confusion matrix results
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Val

### Bug

When running model.val() in Ultralytics versions >= 8.4.112 (possibly earlier), the results for Confusion Matrix and Normalized Confusion Matrix are incoherent.

These are the validation results for a previously trained model on a dataset that has 160 validation images.

Image

Image

The results make little sense - Class Istrice has more false positives than there are images in the validation set, wrong-class predictions are much more frequent than previous versions (see below) and background predictions are ridiculously high across the board.

I have validated the same model on version 8.4.101 and output matches expectations:

Image

Image

This model had been trained on a non-affected version of Ultralytics and exported in NCNN format, but i can confirm this also happens with newly trained models and other formats (can confirm HEF is affected).

### Environment

Ultralytics 8.4.116 🚀 Python-3.10.20 torch-2.13.0+cu130 CUDA:0 (NVIDIA GeForce RTX 4070 SUPER, 12282MiB)
Setup complete ✅ (16 CPUs, 15.6 GB RAM, 112.4/1006.9 GB disk)

OS                     Linux-6.6.87.2-microsoft-standard-WSL2-x86_64-with-glibc2.39
Environment            Linux
Python                 3.10.20
Install                git
Path                   /mnt/c/Users/(redacted)/.venv/lib/python3.10/site-packages/ultralytics
RAM                    15.62 GB
Disk                   112.4/1006.9 GB
CPU                    AMD Ryzen 7 7700X 8-Core Processor
CPU count              16
GPU                    NVIDIA GeForce RTX 4070 SUPER, 12282MiB
GPU count              1
CUDA                   13.0

filelock               ✅ 3.32.2>=3.16.1
numpy                  ✅ 2.2.6>=1.23.0
matplotlib             ✅ 3.10.9>=3.3.0
opencv-python          ✅ 5.0.0.93!=4.13.0.90,>=4.7.0
pillow                 ✅ 12.3.0>=7.1.2
pyyaml                 ✅ 6.0.3>=5.3.1
requests               ✅ 2.34.2>=2.23.0
torch                  ✅ 2.13.0>=1.8.0
torch                  ✅ 2.13.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.28.0>=0.9.0
psutil                 ✅ 7.2.2>=5.8.0
polars                 ✅ 1.43.2>=0.20.0
nvidia-ml-py           ✅ 13.610.43>=12.0.0
ultralytics-thop       ✅ 2.1.6>=2.1.6

### Minimal Reproducible Example

To reproduce, confront the (Normalized) Confusion Matrix output of any model.val() in affected versions (ie. latest/8.4.116) with those from a non-affected version (ie. 8.4.101).
```python3
model = YOLO(model_path, task="detect")
results = model.val(data="./dataset/data.yaml")
print(results.confusion_matrix.matrix)
```

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Rahulbiradar9**
# Validation Confusion Matrix Threshold Update

## Overview
This update improves the readability of validation confusion matrix plots by adjusting the default confidence threshold behavior.

## Problem
- Previous default thresholds:
  - `conf = 0.001` (general)
  - `conf = 0.01` (OBB)
- Result: Plots were cluttered and unreadable.

## Solution
- Added explicit tracking for user-defined thresholds:
  - `self.is_conf_explicit = True` when user sets `conf`.
- Changed default threshold:
  - `conf = 0.25` when not explicitly set.
- Preserves user overrides:
  - Example: `model.val(conf=0.1)` remains valid.

## Implementation
- Modified validation logic to check explicit threshold usage.
- Added dedicated unit test:
  - Location: `tests/test_python.py`
  - Purpose: Verify default threshold behavior and user overrides.

## Benefits
- Clearer, more interpretable confusion matrix plots.
- Maintains flexibility for advanced users.
- Ensures consistent validation visualization across models.

## Usage Example
```python
# Default threshold (0.25 applied automatically)
model.val()

# Explicit threshold override
model.val(conf=0.1)

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Clarifies the confidence threshold used for confusion matrix generation during validation.

### 📊 Key Changes
- Tracks whether the validation confidence threshold was explicitly provided by the user.
- Uses the explicit confidence value for confusion matrix processing, or `0.25` by default when it was not provided.
- Adds tests covering both implicit and explicit confidence configurations.

### 🎯 Purpose & Impact
- Ensures confusion matrix plots use the intended validation confidence behavior.
- Prevents default validator configuration from unintentionally changing confusion matrix results.
- Improves regression coverage for object detection validation and plotting.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Rahulbiradar9, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). This is an automated message, and an Ultralytics engineer will assist with any questions or feedback. Thank you for contributing to Ultralytics! 🚀

**Comment by Rahulbiradar9** [comment]
I have read the CLA Document and I sign the CLA

File: tests/test_python.py [hunk 1]
```diff
@@ -1881,3 +1881,16 @@ def test_semantic_polygon_data():
     model = YOLO("yolo26n-sem.pt")
     model.train(data="coco8-seg.yaml", epochs=1, imgsz=32, close_mosaic=1)
     model.val(data="coco8-seg.yaml")
+
+
+def test_val_confusion_matrix_confidence_threshold():
+    """Test that the confusion matrix uses the correct confidence threshold during validation."""
+    model = YOLO("yolo26n.pt")
+    validator = model._smart_load("validator")(args={"data": "coco8.yaml", "mode": "val", "plots": True})
+    assert not validator.is_conf_explicit
```
Comment (UltralyticsAssistant) [inline tests/test_python.py]: 📝 **LOW**: This test only verifies the new bookkeeping flag; it never exercises `update_metrics()` or observes the `conf` passed to `confusion_matrix.process_batch()`. As a result, reverting the behavior in `detect/val.py` to always use `self.args.conf` would still pass this test. Please mock or invoke the confusion-matrix path and assert that the default call receives `0.25` while an explicit `conf=0.15` receives `0.15`.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The implementation cleanly preserves explicit confidence overrides while using a higher threshold for confusion-matrix plotting. The added test should validate the actual threshold forwarded to the confusion matrix, not only the intermediate flag.

💬 Posted 1 inline comment

**Review by glenn-jocher — APPROVED** [review]
Independently reviewed the exact patched head. The bookkeeping flag and ineffective test are gone; DetectionValidator now owns the raw matrix threshold before BaseValidator fills the metrics default, with direct default/explicit validation and clean focused checks.
